pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -107,7 +107,66 @@ func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitG
 	}
 }
 
-func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, _ int) (err error) {
+func (w *HashAggPartialWorker) newPartialResults(n int, partialResultNum int) [][]aggfuncs.PartialResult {
+	if n > len(w.partialResultsBuffer) {
+		appendNum := n - len(w.partialResultsBuffer)
+		for i := 0; i < appendNum; i++ {
+			w.partialResultsBuffer = append(w.partialResultsBuffer, make([]aggfuncs.PartialResult, partialResultNum))
+		}
+	}
+	return w.partialResultsBuffer[0:0]
+}
+
+func (w *HashAggPartialWorker) expandPartialResults(partialResult [][]aggfuncs.PartialResult, partialResultSize int) [][]aggfuncs.PartialResult {
+	partialResultLen := len(partialResult)
+	if len(w.partialResultsBuffer) < partialResultLen+1 {
+		w.partialResultsBuffer = append(w.partialResultsBuffer, make([]aggfuncs.PartialResult, partialResultSize))
+	}
+	return w.partialResultsBuffer[:partialResultLen+1]
+}
+
+// If the group key has appeared before, reuse the partial result.
+// If the group key has not appeared before, create empty partial results.
+func (w *HashAggPartialWorker) getPartialResultsOfEachRow(_ *stmtctx.StatementContext, groupKey [][]byte, mapper []AggPartialResultMapper, finalConcurrency int) [][]aggfuncs.PartialResult {
+	cntOfGroupKeys := len(groupKey)
+	allMemDelta := int64(0)
+	partialResultNum := w.getPartialResultSliceLenConsiderByteAlign()
+	partialResultsOfEachRow := w.newPartialResults(cntOfGroupKeys, partialResultNum)
+
+	for i := 0; i < cntOfGroupKeys; i++ {
+		finalWorkerIdx := int(murmur3.Sum32(groupKey[i])) % finalConcurrency
+		tmp, ok := mapper[finalWorkerIdx][string(groupKey[i])]
+
+		// This group by key has appeared before, reuse the partial result.
+		if ok {
+			partialResultsOfEachRow = append(partialResultsOfEachRow, tmp)
+			continue
+		}
+
+		// It's the first time that this group by key appeared, create it
+		partialResultsOfEachRow = w.expandPartialResults(partialResultsOfEachRow, partialResultNum)
```

Comment:
expandPartialResults is useless
```suggestion
		partialResultsOfEachRow = append(w.prBuffer, make([]partialResult))
```

Reply:
expandPartialResults is useless

Reply:
> expandPartialResults is useless

okk